### ISSUE TITLE
False positive "await has no effect on the type" diagnostic with mixed generic union

### ISSUE THREAD

**Issue by Andarist** [microsoft/TypeScript#50832]
# Bug Report

### 🔎 Search Terms

await has no effect on type, 80007

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.9.0-dev.20220918#code/IYZwngdgxgBAZgV2gFwJYHsLwgJgDwAqMApgB7LEQAmIMACgE7oC2qIxeAbuqlQHwwAPjG68+ACmAAuGAQCUMAN4AoGDGAB3YKmTqA3MoC+QA)

### 💻 Code

```ts
async function fn2<T extends Promise<void> | void>(a: T) {
  await a; // reports 80007
}
```

### 🙁 Actual behavior

It reports:
> 'await' has no effect on the type of this expression.(80007)

### 🙂 Expected behavior

No problem should be reported here.
